rfc/data-privacy/v2.md

```diff
@@ -0,0 +1,552 @@
+author: Jens Neuse
+---
+
+# Problem
+
+In organizations where data privacy is paramount,
+it's often necessary to obfuscate certain fields in the response to allow developers to debug issues without exposing sensitive information,
+or to export data through APIs for data science or analytics purposes,
+all in compliance with data privacy regulations.
+
+Furthermore, it's important that obfuscation policies are stable,
+ensuring that the data is always obfuscated in the same way,
+allowing developers to rely on the obfuscation behavior.
+
+# Solution
+
+We propose a new feature that allows to obfuscate fields in responses depending on the user's role or the environment.
+The configuration should be done at the Router level as Subgraph developers should not be the single point of control for data privacy.
+The security team should be able to define the rules for obfuscation and the Router should enforce them.
+
+To achieve this, we propose to add a new configuration entry in the Router configuration file that allows to define obfuscation policies.
+Each policy consists of:
+1. An activation rule using expr-lang to determine when the policy applies
+2. Target specification (either YAML-based or expression-based) to determine which fields the policy affects
+3. Obfuscation logic to transform the field values
+
+## Context Objects
+
+For obfuscation policies to work, we need to define context objects that contain all the information necessary to evaluate the expressions:
+
+### Response Field Context Object:
+
+```json
+{
+  "fieldName": "email",
+  "fieldType": "String",
+  "parentType": "User",
+  "value": "user@example.com",
+  "valueDataType": "string"
+}
+```
+
+### Request Context Object:
+
+The "activate" expression has access to the request context object.
+The response context object is defined here: https://cosmo-docs.wundergraph.com/router/configuration/template-expressions
+
+## Configuration Structure
+
+The data privacy configuration consists of:
+
+1. **Transformers**: Reusable obfuscation logic defined using expr-lang
+2. **Policies**: Rules for when and how to apply obfuscation
+3. **Configuration Options**: Options to control behavior and validation
+
+```yaml
+data_privacy:
+  obfuscation:
+    # Configuration options
+    options:
+      undefined_field_behavior: "warn" # Options: "warn", "error", "ignore"
+      validate_transformers: true # Validate transformer return types on startup
+    
+    # Reusable transformers
+    transformers:
+      - name: "email"
+        expression: |
+          repeat("*", len(split(value, "@")[0])) + "@" + split(value, "@")[1]
+      - name: "default"
+        expression: |
+          valueDataType == "boolean" ? false : 
+          valueDataType == "number" ? 0 : 
+          repeat("*", len(value))
+      - name: "isoDate"
+        expression: |
+          date(value, "2006-01-02 15:04:05").Year() + "-**-**"
+    
+    # Obfuscation policies
+    policies:
```

Comment:
are policies supposed to contain mutually exclusive conditions in their `activate` blocks? can multiple policies be applied one after another?

Reply:
Multiple policies can be active, yes, but if one active policy matches a field, we will not evaluate further. Is this in line with your expectations or are you thinking of having multiple policies active?

Reply:
we don't need multiple policies to be active. "first policy wins" is good enough 👍